src/compute/src/render/mod.rs

```diff
@@ -919,3 +920,57 @@ impl<T: Timestamp + Lattice> RenderTimestamp for Product<mz_repr::Timestamp, T>
         Product::new(self.outer.saturating_sub(1), self.inner.clone())
     }
 }
+
+/// Used to make possibly-validating code generic: think of this as a kind of `MaybeResult`,
+/// specialized for use in compute.  The contract is that the error constructor for non-validating
+/// implementors will never be called.
```

Comment:
Not sure if that improves things, but you could encode this contract in the type system by having, e.g. an `into_error` trait method that returns an `Option<&dyn Fn(E) -> Self>` (or whatever actually type-checks). Then non-validating types could return `None` here, so there is no error constructor that could be invoked for them.

Maybe that makes things more complicated than they need to be...

Reply:
That's an interesting idea.  I'll play with it a bit.

Reply:
This works nicely in fact.  Great idea.  I would like to think a little bit about the name, since this signals both that we're validating, and gives us the error constructor.